hub/core/storage/gcs.py

```diff
@@ -0,0 +1,277 @@
+                If None, will try sequence of methods.
+
+        Raises:
+            AttributeError: If method is invalid.
+        """
+        if method not in [
+            "google_default",
+            "cache",
+            "anon",
+            "browser",
+            None,
+        ]:
+            self._connect_token(method)
+        elif method is None:
+            for meth in ["google_default", "cache", "anon"]:
+                self.connect(method=meth)
+                break
+        elif isinstance(method, str):
+            self.__getattribute__("_connect_" + method)()
+        else:
+            raise AttributeError(f"Invalid method: {method}")
+
+
+class GCSProvider(StorageProvider):
+    """Provider class for using GC storage."""
+
+    def __init__(self, root: str, token: Union[str, Dict] = None, project: str = None):
+        """Initializes the GCSProvider
+
+        Example:
+            gcs_provider = GCSProvider("gcs://my-bucket/gcs_ds")
+
+        Args:
+            root (str): The root of the provider. All read/write request keys will be appended to root.
+            token (str/Dict): GCP token, used for fetching credentials for storage).
+                Can be a path to the credentials file, actual credential dictionary or one of the folowing:
+                `google_default`: Tries to load default credentials for the specified project.
+                `cache`: Retrieves the previously used credentials from cache if exist.
+                `anon`: Sets credentials=None
+                `browser`: Generates and stores new token file using cli.
+            project (str): Name of the project from gcloud.
+        """
+        self.root = root
+        self.token: Union[str, Dict, None] = token
+        self.project = project
+        self.missing_exceptions = (
+            FileNotFoundError,
+            IsADirectoryError,
+            NotADirectoryError,
+            AttributeError,
+        )
+        self._initialize_provider()
+
+    def _initialize_provider(self):
+        self._set_bucket_and_path()
+        if not self.token:
+            self.token = None
+        self.scoped_credentials = GCloudCredentials(self.token, project=self.project)
+        self.retry = retry.Retry(deadline=60)
+        client = storage.Client(credentials=self.scoped_credentials.credentials)
+        self.client_bucket = client.get_bucket(self.bucket)
+
+    def _reinitialize_provider(self):
+        client = storage.Client(credentials=self.scoped_credentials.credentials)
+        self.client_bucket = client.get_bucket(self.bucket)
+
+    def _set_bucket_and_path(self):
+        root = self.root.replace("gcp://", "").replace("gcs://", "")
+        self.bucket = root.split("/")[0]
+        self.path = root
+        if not self.path.endswith("/"):
+            self.path += "/"
+
+    def _get_path_from_key(self, key):
+        return posixpath.join(self.path, key)
+
+    def _list_keys(self):
```

Comment:
I've renamed the method to _all_keys in main which returns a set, could you update this too?